### ISSUE TITLE
AFHTTPBodyPart: synchronous use of main queue may lead to deadlocks

### ISSUE THREAD

**Issue by snake-scaly** [AFNetworking/AFNetworking#2481]
The method, `-[AFHTTPBodyPart transitionToNextPhase]` contains code:

``` ObjC
    if (![[NSThread currentThread] isMainThread]) {
        dispatch_sync(dispatch_get_main_queue(), ^{
            [self transitionToNextPhase];
        });
        return YES;
    }
```

This code _did_ cause deadlocks for me. It happened like this:
1. A background thread calls a method on a large stateful object
2. The large stateful object synchronizes on its state mutex
3. The large stateful object attempts to post a multipart request
4. Meanwhile the main thread calls another method on that large stateful object
5. The large stateful object attempts to synchronize on its state mutex, and the main thread blocks
6. The `transitionToNextPhase` method attempts to synchronously execute on the main thread and deadlocks

This dependency on the main thread is unexpected. Can it be removed?

**Comment by foby** [comment]
I have the same problem: my input stream works synchronously and thus needs to run on a background thread.

What is the reason for `transitionToNextPhase` is being executed on the main queue?

**Comment by kaybutter** [comment]
I have the same problem. I'm trying to test a file upload in a unit test. The test waits for the request to finish by blocking the main thread and request can never finish because it attempts to dispatch on the main thread itself. :(

**Comment by qiulang** [comment]
Because we should never access a scheduled stream from a thread different than the one owning the stream’s run loop a more valid question would be can we use a dedicated thread for AFHTTPBodyPart operation ?
For example can we use dispatch_async(/* a dedicated queue */ ... ) instead of dispatch_sync(dispatch_get_main_queue() ...) ?

**Comment by snake-scaly** [comment]
@qiulang Please be aware that a queue is quite different from a thread. Each job on a queue is executed on a random thread from a thread pool, even if that queue is sequential. The only queue that guarantees execution of all jobs on the same thread is the main queue.

**Comment by qiulang** [comment]
Hi the code really matter is this
[self.inputStream scheduleInRunLoop:[NSRunLoop currentRunLoop]  forMode:NSRunLoopCommonModes];  

and self.inputStream is created the first time accessed
So if you said the thread creates self.inputStream can be different from the threads run scheduleInRunLoop even if we put them in dispatch_async then we are out of option. But I don't think that is the case.

**Comment by snake-scaly** [comment]
@qiulang Yes I've seen that code you mention and it puzzled me. The only reason to schedule an input stream on a run loop is to be able to receive data from it asynchronously. Or, to put it differently, scheduling only makes sense if you implement a stream delegate. AFNetworking does not do that. Instead it actively polls the stream until all data is consumed. You can throw that `scheduleInRunLoop` line away and nothing will change.

Another consideration is, it seems to be safe to create an input stream in one thread and then read in another. That's what happens when you schedule a stream on a different thread's run loop after all. It even should be safe to perform separate reads on different threads if you guarantee that you never read in parallel. The latter is only my guess though. I couldn't find any official documentation on NSStream thread safety.

To summarize, I think both `dispatch_sync()` and `[scheduleInRunLoop]` can be removed from the code without any negative consequences. However I wonder why it was written like this originally. There could have been a reason.

**Comment by qiulang** [comment]
Let's just hope the author can explain it. 

But some of your statements were wrong I guess, because apple document did said "Before you open the stream to begin the streaming of data, send a scheduleInRunLoop:forMode: message to the stream object to schedule it to receive stream events on a run loop... If streaming is taking place on another thread, be sure to schedule the stream object on that thread’s run loop. "

https://developer.apple.com/library/mac/documentation/Cocoa/Conceptual/Streams/Articles/ReadingInputStreams.html#//apple_ref/doc/uid/20002273-BCIJHAGD

**Comment by kcharwood** [comment]
@snake-scaly would you be able to create a unit test that demonstrates the failure?

**Comment by qiulang** [comment]
There is an easy way to reproduce the issue and we actually have code like that :( 

1 create a dedicated queue(let's call it resyncQueue) for AFN and set operation.completionQueue = resyncQueue;  

2 In the success block call that POST method. So now resyncQueue is waiting for main queue.

3. Somewhere in main queue, it also is waitsing for resyncQueue for other data (I don't think this is unusual) and now you have deadlock!

**Comment by appleguy** [comment]
@kcharwood hey there - any chance you know how this was resolved in a later version of AF, or if it was?  Pinterest is using 2.6.3, and this is our #1 stability problem by far.

As is the case for many apps, Pinterest requires the ability to occasionally block the main thread on the completion of concurrent background operations.  This means than any dispatch_sync calls from background to main can deadlock.

In general, dispatch_sync is not safe for framework code if it uses a serial queue that is accessible outside the framework.  AsyncDisplayKit has managed to get away without using even one dispatch_sync, and I'm pretty confident AFNetworking can too.

Anyway, if there is a patch available for this we can apply locally, it would make a huge difference for us.  cc @rahul-malik, @johnepinterest, @garrettmoon

**Comment by tclementdev** [comment]
I'm also seeing a deadlock due to reentrancy:

    Thread 11 Crashed:: Dispatch queue: com.apple.main-thread
        _dispatch_barrier_sync_f_slow + 675
        -[AFHTTPBodyPart transitionToNextPhase] + 232
        __39-[AFHTTPBodyPart transitionToNextPhase]_block_invoke + 38
        _dispatch_client_callout + 8
        _dispatch_barrier_sync_f_invoke + 83
        -[AFHTTPBodyPart transitionToNextPhase] + 232

Note `Thread 11` and `com.apple.main-thread`.

It seems we cannot make the asumption anymore that `dispatch_sync(dispatch_get_main_queue(), ...)` actually executes on the main thread. This may be new on Sierra, I am not sure.

I would also agree with @appleguy that using the main queue here is not appropriate since this queue may be used by unknown code outside of AFNetworking.

Using a private serial queue seems more appropriate.

**Comment by appleguy** [comment]
Eager to learn any updates here...AF had served us very well for a long
time, but it is specifically this issue that made it essential for
Pinterest to port off of the framework.

We found that a lot of network performance can be gained by not waiting on
a trip through the main thread just to switch states.  This often gets
stuck behind larger operations like a collection layout or first
transaction commit after pushing a view controller, and can add a few
hundred milliseconds in avoidable latency. I would be happy to continue to
recommend it if this could be fixed.

On Sep 1, 2016 11:11 AM, "tclementdev" <notifications@github.com> wrote:

I'm also seeing a deadlock due to reentrancy:

Thread 11 Crashed:: Dispatch queue: com.apple.main-thread
    _dispatch_barrier_sync_f_slow + 675
    -[AFHTTPBodyPart transitionToNextPhase] + 232
    __39-[AFHTTPBodyPart transitionToNextPhase]_block_invoke + 38
    _dispatch_client_callout + 8
    _dispatch_barrier_sync_f_invoke + 83
    -[AFHTTPBodyPart transitionToNextPhase] + 232

Note Thread 11 and com.apple.main-thread.

It seems we cannot make the asumption anymore that
dispatch_sync(dispatch_get_main_queue(), ...) actually executes on the main
thread. This may be new on Sierra, I am not sure.

I would also agree with @appleguy <https://github.com/appleguy> that using
the main queue here is not appropriate since this queue may be used by
unknown code outside of AFNetworking.

Using a private serial queue seems more appropriate.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/AFNetworking/AFNetworking/issues/2481#issuecomment-244110186>,
or mute the thread

**Comment by mschneider** [comment]
> Using a private serial queue seems more appropriate.

This actually worked well for me to resolve the issue in our project

**Comment by tclementdev** [comment]
I don't think that's the right fix here. This code (some part of it at least) relies on being on a thread that runs a run loop. You won't have that in a libdispatch thread (and the NSInputStream won't work as expected).

**Comment by tewha** [comment]
Guys, what does `transitionToNextPhase` actually do? It looks like an elaborate no-op or deadlock to me.

**Comment by wangjufan** [comment]
Hi everyone ,  in our project we with replace the sys codes with gcd semaphore .


typedef enum {
    AFEncapsulationBoundaryPhase = 1,
    AFHeaderPhase                = 2,
    AFBodyPhase                  = 3,
    AFFinalBoundaryPhase         = 4,
} AFHTTPBodyPartReadPhase;

@interface AFHTTPBodyPart () <NSCopying> {
    AFHTTPBodyPartReadPhase _phase;
    dispatch_semaphore_t _phaseSemaphore;
    NSInputStream *_inputStream;
    unsigned long long _phaseReadOffset;
}

- (BOOL)transitionToNextPhase;
- (NSInteger)readData:(NSData *)data
           intoBuffer:(uint8_t *)buffer
            maxLength:(NSUInteger)length;
@end

@implementation AFHTTPBodyPart

- (instancetype)init {
    self = [super init];
    if (!self) {
        return nil;
    }
    
    _phaseSemaphore = dispatch_semaphore_create(1);
    dispatch_semaphore_wait(_phaseSemaphore, DISPATCH_TIME_FOREVER);
    [self transitionToNextPhase];
    dispatch_semaphore_signal(_phaseSemaphore);
    
    return self;
}

- (void)dealloc {
    if (_inputStream) {
        [_inputStream close];
        _inputStream = nil;
    }
}

- (NSInputStream *)inputStream {
    if (!_inputStream) {
        if ([self.body isKindOfClass:[NSData class]]) {
            _inputStream = [NSInputStream inputStreamWithData:self.body];
        } else if ([self.body isKindOfClass:[NSURL class]]) {
            _inputStream = [NSInputStream inputStreamWithURL:self.body];
        } else if ([self.body isKindOfClass:[NSInputStream class]]) {
            _inputStream = self.body;
        } else {
            _inputStream = [NSInputStream inputStreamWithData:[NSData data]];
        }
    }
    
    return _inputStream;
}

- (NSString *)stringForHeaders {
    NSMutableString *headerString = [NSMutableString string];
    for (NSString *field in [self.headers allKeys]) {
        [headerString appendString:[NSString stringWithFormat:@"%@: %@%@", field, [self.headers valueForKey:field], kAFMultipartFormCRLF]];
    }
    [headerString appendString:kAFMultipartFormCRLF];
    
    return [NSString stringWithString:headerString];
}

- (unsigned long long)contentLength {
    unsigned long long length = 0;
    
    NSData *encapsulationBoundaryData = [([self hasInitialBoundary] ? AFMultipartFormInitialBoundary(self.boundary) : AFMultipartFormEncapsulationBoundary(self.boundary)) dataUsingEncoding:self.stringEncoding];
    length += [encapsulationBoundaryData length];
    
    NSData *headersData = [[self stringForHeaders] dataUsingEncoding:self.stringEncoding];
    length += [headersData length];
    
    length += _bodyContentLength;
    
    NSData *closingBoundaryData = ([self hasFinalBoundary] ? [AFMultipartFormFinalBoundary(self.boundary) dataUsingEncoding:self.stringEncoding] : [NSData data]);
    length += [closingBoundaryData length];
    
    return length;
}

- (BOOL)hasBytesAvailable {
    // Allows `read:maxLength:` to be called again if `AFMultipartFormFinalBoundary` doesn't fit into the available buffer
    if (_phase == AFFinalBoundaryPhase) {
        return YES;
    }
    
    switch (self.inputStream.streamStatus) {
        case NSStreamStatusNotOpen:
        case NSStreamStatusOpening:
        case NSStreamStatusOpen:
        case NSStreamStatusReading:
        case NSStreamStatusWriting:
            return YES;
        case NSStreamStatusAtEnd:
        case NSStreamStatusClosed:
        case NSStreamStatusError:
        default:
            return NO;
    }
}

- (NSInteger)read:(uint8_t *)buffer
        maxLength:(NSUInteger)length
{
    NSInteger totalNumberOfBytesRead = 0;
    
    if (_phase == AFEncapsulationBoundaryPhase) {
        NSData *encapsulationBoundaryData = [([self hasInitialBoundary] ? AFMultipartFormInitialBoundary(self.boundary) : AFMultipartFormEncapsulationBoundary(self.boundary)) dataUsingEncoding:self.stringEncoding];
        totalNumberOfBytesRead += [self readData:encapsulationBoundaryData intoBuffer:&buffer[totalNumberOfBytesRead] maxLength:(length - (NSUInteger)totalNumberOfBytesRead)];
    }
    
    if (_phase == AFHeaderPhase) {
        NSData *headersData = [[self stringForHeaders] dataUsingEncoding:self.stringEncoding];
        totalNumberOfBytesRead += [self readData:headersData intoBuffer:&buffer[totalNumberOfBytesRead] maxLength:(length - (NSUInteger)totalNumberOfBytesRead)];
    }
    
    if (_phase == AFBodyPhase) {
        NSInteger numberOfBytesRead = 0;
        
        numberOfBytesRead = [self.inputStream read:&buffer[totalNumberOfBytesRead] maxLength:(length - (NSUInteger)totalNumberOfBytesRead)];
        if (numberOfBytesRead == -1) {
            return -1;
        } else {
            totalNumberOfBytesRead += numberOfBytesRead;
            
            if ([self.inputStream streamStatus] >= NSStreamStatusAtEnd) {
                dispatch_semaphore_wait(_phaseSemaphore, DISPATCH_TIME_FOREVER);
                [self transitionToNextPhase];
                dispatch_semaphore_signal(_phaseSemaphore);
            }
        }
    }
    
    if (_phase == AFFinalBoundaryPhase) {
        NSData *closingBoundaryData = ([self hasFinalBoundary] ? [AFMultipartFormFinalBoundary(self.boundary) dataUsingEncoding:self.stringEncoding] : [NSData data]);
        totalNumberOfBytesRead += [self readData:closingBoundaryData intoBuffer:&buffer[totalNumberOfBytesRead] maxLength:(length - (NSUInteger)totalNumberOfBytesRead)];
    }
    
    return totalNumberOfBytesRead;
}

- (NSInteger)readData:(NSData *)data
           intoBuffer:(uint8_t *)buffer
            maxLength:(NSUInteger)length
{
    NSRange range = NSMakeRange((NSUInteger)_phaseReadOffset, MIN([data length] - ((NSUInteger)_phaseReadOffset), length));
    [data getBytes:buffer range:range];
    
    _phaseReadOffset += range.length;
    
    if (((NSUInteger)_phaseReadOffset) >= [data length]) {
        dispatch_semaphore_wait(_phaseSemaphore, DISPATCH_TIME_FOREVER);
        [self transitionToNextPhase];
        dispatch_semaphore_signal(_phaseSemaphore);
    }
    
    return (NSInteger)range.length;
}

- (BOOL)transitionToNextPhase {
    
    //    if (![[NSThread currentThread] isMainThread]) {
    //        dispatch_sync(dispatch_get_main_queue(), ^{
    //            [self transitionToNextPhase];
    //        });
    //        return YES;
    //    }
    
    switch (_phase) {
        case AFEncapsulationBoundaryPhase:
            _phase = AFHeaderPhase;
            break;
        case AFHeaderPhase:
            [self.inputStream scheduleInRunLoop:[NSRunLoop currentRunLoop] forMode:NSRunLoopCommonModes];
            [self.inputStream open];
            _phase = AFBodyPhase;
            break;
        case AFBodyPhase:
            [self.inputStream close];
            _phase = AFFinalBoundaryPhase;
            break;
        case AFFinalBoundaryPhase:
        default:
            _phase = AFEncapsulationBoundaryPhase;
            break;
    }
    _phaseReadOffset = 0;
    
    return YES;
}

#pragma mark - NSCopying

- (instancetype)copyWithZone:(NSZone *)zone {
    AFHTTPBodyPart *bodyPart = [[[self class] allocWithZone:zone] init];
    
    bodyPart.stringEncoding = self.stringEncoding;
    bodyPart.headers = self.headers;
    bodyPart.bodyContentLength = self.bodyContentLength;
    bodyPart.body = self.body;
    bodyPart.boundary = self.boundary;
    
    return bodyPart;
}

@end

**Comment by tclementdev** [comment]
That doesn't seem right either. Consider this line:

    [self.inputStream scheduleInRunLoop:[NSRunLoop currentRunLoop] forMode:NSRunLoopCommonModes];

You need the thread to run the run loop for this to work. What if `-transitionToNextPhase` is called from a secondary/libdispatch thread?

**Comment by wangjufan** [comment]
Yes , that's rigth !  
But you should never attempt to access a scheduled input stream from a different thread that run on a different runloop  @tclementdev  . 
So we should run the send the close message from scheduled runloop to stream object, but directly use close  in dealloc method .
Please see Stream Programming Guide .

**Comment by wangjufan** [comment]
There are two ways to make use of NSInputStream , polling vs run loop scheduling , while we use scheduling without setting the delegate of the stream and read in data with polling  . 
This is a big mistake .   
I have removed [self.inputStream scheduleInRunLoop:[NSRunLoop currentRunLoop] forMode:NSRunLoopCommonModes]; , it  works well .

**Comment by tclementdev** [comment]
Well if there is no need for a run loop, then that solves the problem.

Note that you shouldn't use `dispatch_semaphore_t` for mutual exclusion, just use a lock. For example a private dispatch queue with `dispatch_sync()` calls. Generally speaking, dispatch semaphores should be avoided as they do not play well with priorities and QOS. The libdispatch cannot reason about when the semaphore will be signaled (while the `dispatch_sync()` block is well defined).

**Comment by wangjufan** [comment]
Yes , you are right . Using dispatch_semaphore_t in this context is an mistake . 
Keeping the execution order is an intrinsic requirement in the situation . 
So reentrant lock or serial queue works well .

**Comment by wsndy163** [comment]
@wangjufan How did you solve it in the end?
